### PR TITLE
#2275 - Org Model view only message and tooltip fixes - [AB]

### PR DESCRIPTION
**Pull request by CocoByte**
## Ticket [2275](https://github.com/cisagov/manage.get.gov/issues/2775)

Resolves #[2275](https://github.com/cisagov/manage.get.gov/issues/2775)

## Changes

- Updates the message that appears on the Domain page if a user can only view a domain, but not manage it. See the text in [this document](https://docs.google.com/document/d/1u-K1yQKb16H3AWj33mH2e1LJme6Ay9dNXMG8isiOXWk/edit#heading=h.tsr9qh4ofhbi).
- Fixes tooltips in the Domains table so that their content doesn't get clipped by parent containers (this is accomplished in two ways -- first, by adjusting wrapping logic so we can specify max widths to be less than the viewport width.  Secondly, by adjusting the positioning logic so the tooltip can show above its parent container(s).  Between these two mechanisms, we should be able to easily customize the tooltip to design needs)
- The above tooltip fix also adds a child element "usa-tooltip__content" to the tooltip.  This was needed for making the viewport adjustments for wrapping behavior.  But a handy side-effect of this is if design ever wants a fancy tooltip, we can absolutely make it as fancy as we want now by modifying this child element.

## Context for reviewers

NOTE @rachidatecs -- I remember you mentioning that it is preferable to use `<span>` within the tooltip.  However, using a `<span>` instead of a `<div>` (or other containers) causes conflicts in positioning.  `<span>` is getting handled differently and while I can see how using `<span>` would be the more minimal solution (especially for text content), `<div>` I think is more extensible/robust.  Perhaps I could add `<span>` within the `<div>` to keep text within appropriate containers...but this would *assume* purely text-based content (what if we wanted HTML instead?).  And since making assumptions in code is also breaking best practices, I opted to just use a `<div>` for simplicity and extensibility.

## Code Review Verification Steps

- [ ] Login as a user that can only view a domain, but not manage it.  Navigate to the domain page for that domain, and verify that the text is updated to "You don't have access to manage < domain >. If you need to make updates, contact one of the listed domain managers."
- [ ] Go back to the Domains table and verify that a various window widths, the tooltips do not have their content clipped.  (See image below, which shows a tooltip modified with lots of text to demonstrate wrapping)

image

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

**Comment by Katherine-Osos** [comment]
@CocoByte Since we won't have "member permissions" rolled out before people see this message ... can you remove the third sentence for now? (We'll need to adjust the message again after admins are incorporated into the system.) 

Therefore, it should read: 

> You don't have access to manage < domain >. If you need to make updates, contact one of the listed domain managers.

**Comment by rachidatecs** [comment]
NOTE: You're linking to another PR instead of the ticket at the top of your PR description

File: src/registrar/assets/js/uswds-edited.js [hunk 1]
```diff
@@ -6128,7 +6158,13 @@ const setUpAttributes = tooltipTrigger => {
   tooltipBody.setAttribute("aria-hidden", "true");
 
   // place the text in the tooltip
-  tooltipBody.textContent = tooltipContent;
+  // DOTGOV: nest the text element to allow us creater control over width and wrapping behavior
```
Comment (rachidatecs) [inline src/registrar/assets/js/uswds-edited.js:6161]:
```suggestion
  // DOTGOV: nest the text element to allow us greater control over width and wrapping behavior
```

File: src/registrar/assets/js/uswds-edited.js [hunk 2]
```diff
@@ -5949,8 +5966,10 @@ const showToolTip = (tooltipBody, tooltipTrigger, position) => {
     const topMargin = calculateMarginOffset("top", e.offsetHeight, tooltipTrigger);
     const leftMargin = calculateMarginOffset("left", e.offsetWidth, tooltipTrigger);
     setPositionClass("top");
-    e.style.left = `50%`; // center the element
-    e.style.top = `-${TRIANGLE_SIZE}px`; // consider the pseudo element
+
+    // DOTGOV: updated logic for position:fixed
```
Comment (rachidatecs) [inline src/registrar/assets/js/uswds-edited.js:5970]: BLOCKING: For all changes in this file, I suggest you wrap them (even if it's one line) with DOTGOV EDIT and END DOTGOV EDIT comments. Leave the original USWDS code within these blocks, commented out.

We want to make it as easy as possible for devs who do not have context to update USWDS. To do so, they have to compile new releases and then go in and redo our edits onto the compiled code. I think my suggestion takes out some cognitive effort from this task.

File: src/registrar/assets/sass/_theme/_tooltips.scss [hunk 3]
```diff
@@ -28,3 +28,46 @@
...
+    width: 50vw;
+    text-wrap: wrap;
+    text-align: center;
+    font-size: inherit; //inherit tooltip fontsize of .93rem
+    max-width: fit-content;
+    @include at-media('tablet') {
+      width: 70vw;
+    }
+  }
+}
```
Comment (rachidatecs) [inline src/registrar/assets/sass/_theme/_tooltips.scss:73]: NITPICK: endline

File: src/registrar/assets/js/uswds-edited.js [hunk 4]
```diff
@@ -6128,7 +6158,13 @@ const setUpAttributes = tooltipTrigger => {
   tooltipBody.setAttribute("aria-hidden", "true");
 
   // place the text in the tooltip
-  tooltipBody.textContent = tooltipContent;
+  // DOTGOV: nest the text element to allow us creater control over width and wrapping behavior
+   tooltipBody.innerHTML = `
+    <div class="usa-tooltip__content">
```
Comment (rachidatecs) [inline src/registrar/assets/js/uswds-edited.js:6163]: The div inside a span does not seem to trigger accessibility check errors, but it goes counter to HTML specifications. A div is a block level container, a span is an inline-level container span[https://developer.mozilla.org/en-US/docs/Web/HTML/Element/span]

CSS-wise, you can easily make a span behave like a div in the layout by setting display: block on it. 

Q: Is there something else that's keeping you from using a span with a display block style on?

**Comment by rachidatecs** [comment]
NOTE: Code is no longer on NL. Deploy it somewhere else for the design review.

File: src/registrar/assets/sass/_theme/_tooltips.scss [hunk 5]
```diff
@@ -28,3 +28,46 @@
...
+  /*
+  Add width and wrapping to tooltip content in order to confine it to a smaller viewport window.
+  */
+  .usa-tooltip__content {
+    width: 50vw;
+    text-wrap: wrap;
+    text-align: center;
+    font-size: inherit; //inherit tooltip fontsize of .93rem
+    max-width: fit-content;
+    @include at-media('tablet') {
```
Comment (rachidatecs) [inline src/registrar/assets/sass/_theme/_tooltips.scss:69]: Suggest changing this media query from tablet to desktop. It'll just be a bit more stable. See screenshot:

Screenshot 2024-10-09 at 7 31 07 PM

**Comment by rachidatecs** [comment]
Fantastic work.

File: src/registrar/assets/js/uswds-edited.js (hunk 2 above)
Comment (CocoByte) [reply src/registrar/assets/js/uswds-edited.js:5970]: Totally agree.  I like this suggestion.

File: src/registrar/assets/js/uswds-edited.js (hunk 4 above)
Comment (CocoByte) [reply src/registrar/assets/js/uswds-edited.js:6163]: I see your point, and I get nesting divs inside spans is generally a no-no.   However, the element is set to display:block already when it becomes visible (.usa-tooltip__body.is-set), yet there is something else going on that positions the tooltip incorrectly above the tooltip icon when we use `<span>`.  We might want to huddle on this.

**Review by rachidatecs — APPROVED** [review]

**Pull request merged**